### PR TITLE
Improvements and advice

### PR DESCRIPTION
**Pull request by viggy96**
Looking at your code, one thing that hit me was that you guys had three subsystems, for a robot that essentially just has a shooter. While I technically understand that your Zerobot shooter could be considered three separate mechanisms, they are together effectively one unit. And putting them together into a single subsystem makes integrating the actions between all three motors much easier.

Also, you can use a static class or enum to store your shooter states for various things (intaking from source, shooting from podium, shooting from subwoofer, etc). 

In `Constants.java`, I personally like to have separate static classes for my hardware and other settings. So if you notice in PH2024, you'll see we have `Drive` and `DriveHardware`, and similar for other subsystems. 

For your indexer, you shouldn't need to use PID on that, it is not super critical that that motor be at a specific RPM at any point. But I imagine you guys did it as practice for learning PID, which is fine. PurpleLib does enable voltage compensation for you on the Sparks, so even duty cycle mode is somewhat accurate. Also PurpleLib already restores factory defaults and burns flash when you call `initializeSparkPID`, so no need to call those methods on your own. 

Regarding code style, its important to have your constructor as the first thing in your class, and have correct indentation.
Formalizing to certain rules helps with this. The WPILib project generator appears to default to use indenting with spaces, with each tabstop being 2 spaces. Enabling "trim trailing whitespace" in VSCode helps a bit too, so extra whitespace between programmers doesn't get picked up as a code change.

Also, for your motion PID, you can take advantage of PurpleLib's "smooth motion" feature, which handles the creation of a trapezoidal motion profile for you, runs the profile by sending a constant stream of position setpoints to the Spark in regular "ControlType.kPosition" mode (DO NOT USE SMARTMOTION!), and stopping when the motion is complete. All this is handled in the call to `periodic()` for each motor controller. To use this feature, simply call `m_spark.smoothMotion()` with the sensor position you want to go to, your trapezoidal motion constraint, and optionally a lambda function or supplier that provides a feed forward voltage value as a double. 

Another thing regarding feed forward is, I haven't found it necessary to use WPILib feed forward for velocity, I just set kF equal to `1 / motorKind.getMaxRPM()` if your units are RPM, and the Spark will take care of the rest. 
For tangential velocity, kF should be something like `1 / ((Spark.MotorKind.NEO_VORTEX.getMaxRPM() / 60) * (FLYWHEEL_DIAMETER.in(Units.Meters) * Math.PI / gearRatio))`.

You guys should also update your PurpleLib fork when you can, as we have provided improvements to many things, including encoder velocity filtering, FFConstants object, etc.

I overall tried not to touch what you guys had already done, but rather provide another example of how to do things, in the currently recommended way by WPILib. Somewhat enabled by having all three motors in the same subsystem, it becomes somewhat trivial to have "command factory" methods in your subsystem that composes the various methods of that subsystem into an action, and returns a command. With this structure, most methods should be private, except command providers and status getters. I didn't add any subsystem level output logging (using `Logger.recordOutput()`), but feel free to add some.

I also updated your `DriveSubsystem` to pretty much the latest of what we've got in our PH2024 repository, and taken out the weirdness that we had but you guys (unfortunately) copied. Sorry, that one is on us. We originally had to turn off the automatic path flipping for PathPlanner, because we use our own pathfinding system that runs on our coprocessor, and because of that we had extra logic in our `AutoTrajectory` class that selected the blue or red auto, and we had to manually create blue and red versions of each auto. However, since then we have restructured our code, and we have turned PathPlannerLib's path flipping off just for our pathfinder, allowing us to turn it on for regular autos. Now you should be able to use the PathPlanner GUI to create autos, with named commands and all the rest of it. Look at [pathplanner.dev](https://pathplanner.dev/home.html) for more info. Feel free to continue to use our PH2024 repo as reference.

We also noticed that we had weirdness when we reset the encoders in our `resetPose()` method, it messes up our autos. So we took that out, and I took it out for you guys as well. mjansen (the PathPlanner dev) swears that it resets the robot's pose on its own so long as an initial pose is set in the PathPlanner GUI, but we've been having trouble with that, so we still reset the pose initially on our own, which you can already see.

I did like that you guys had decided on motor inversions based on the game piece motion. As in shooting out is positive, intaking is negative. Doing things this way is better than just spreading negative signs all over the place. Once you get to the real robot, and you find a motor is spinning the wrong way for shooting/intaking/whatever, adding a call to `m_spark.setInverted(true)` is better than remembering to put a negative in a million places.

So to summarize:

- Have clean code with consistent code style
- Structure things in a way that makes the code modular, and easy to maintain and understand.
- Combine hardware that is logically one mechanism to simplify the code
- Take advantage of Spark PID, no need for external kF for velocity PID, and utilise PurpleLib smooth motion for position control
- Utilize an enum or static class to hold system states for subsystems that need it
- Use "command factory" methods to generate commands in your subsystems
- Enjoy using PurpleLib and ask questions!

### CODE REVIEW

**Comment by viggy96** [comment]
We've also got shooter angle mechanism simulation in our repo too. We haven't gotten to actually simulate shooting game pieces.
